src/Common/OvercommitTracker.cpp

```diff
@@ -36,42 +39,86 @@ bool OvercommitTracker::needToStopQuery(MemoryTracker * tracker)
         return true;
 
     pickQueryToExclude();
-    assert(cancelation_state == QueryCancelationState::RUNNING);
+    assert(cancellation_state != QueryCancellationState::NONE);
     global_lock.unlock();
 
     // If no query was chosen we need to stop current query.
     // This may happen if no soft limit is set.
     if (picked_tracker == nullptr)
     {
-        cancelation_state = QueryCancelationState::NONE;
+        // Here state can not be RUNNING, because it requires
+        // picked_tracker to be not null pointer.
+        assert(cancellation_state == QueryCancellationState::SELECTED);
+        cancellation_state = QueryCancellationState::NONE;
         return true;
     }
     if (picked_tracker == tracker)
+    {
+        // Query of the provided as an argument memory tracker was chosen.
+        // It may happen even when current state is RUNNING, because
+        // ThreadStatus::~ThreadStatus may call MemoryTracker::alloc.
+        cancellation_state = QueryCancellationState::RUNNING;
         return true;
-    bool timeout = !cv.wait_for(lk, max_wait_time, [this]()
+    }
+
+    allow_release = true;
+
+    required_memory += amount;
+    required_per_thread[tracker] = amount;
+    bool timeout = !cv.wait_for(lk, max_wait_time, [this, tracker]()
     {
-        return cancelation_state == QueryCancelationState::NONE;
+        return required_per_thread[tracker] == 0 || cancellation_state == QueryCancellationState::NONE;
     });
-    if (timeout)
-        LOG_DEBUG(getLogger(), "Need to stop query because reached waiting timeout");
-    else
-        LOG_DEBUG(getLogger(), "Memory freed within timeout");
-    return timeout;
+    LOG_DEBUG(getLogger(), "Memory was{} freed within timeout", (timeout ? " not" : ""));
+
+    required_memory -= amount;
+    Int64 still_need = required_per_thread[tracker]; // If enough memory is freed it will be 0
+    required_per_thread.erase(tracker);
+
+    // If threads where not released since last call of this method,
+    // we can release them now.
+    if (allow_release && required_memory <= freed_memory && still_need != 0)
```

Comment:
This check is correct. If here `still_need != 0`, that means we need to stop this query, because there is no enough free memory.
The purpose of this check is to allow other threads to continue execution if after cancelling of the current query it's enough memory for rest queries.

I didn't understand question about `releaseThreads`/`reset`.

Reply:
>The purpose of this check is to allow other threads to continue execution if after cancelling of the current query it's enough memory for rest queries.

Ok, I see thanks.

>I didn't understand question about releaseThreads/reset.

I thought that the purpose of this check is to allow execution of current query too, hence the question.